pkg/operator/defragcontroller/defragcontroller.go

```diff
@@ -0,0 +1,188 @@
+	operatorv1 "github.com/openshift/api/operator/v1"
+	configv1listers "github.com/openshift/client-go/config/listers/config/v1"
+	"github.com/openshift/cluster-etcd-operator/pkg/etcdcli"
+	"github.com/openshift/cluster-etcd-operator/pkg/operator/ceohelpers"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	"go.etcd.io/etcd/clientv3"
+	"go.etcd.io/etcd/etcdserver/etcdserverpb"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog/v2"
+)
+
+const (
+	minDefragBytes          int64   = 100 * 1024 * 1024 // 100MB
+	maxFragmentedPercentage float64 = 45
+	waitDuration                    = 2 * time.Second
+	timeoutDuration                 = 30 * time.Second
+)
+
+// DefragController observes the operand state file for fragmentation
+type DefragController struct {
+	operatorClient       v1helpers.OperatorClient
+	etcdClient           etcdcli.EtcdClient
+	infrastructureLister configv1listers.InfrastructureLister
+}
+
+func NewDefragController(
+	operatorClient v1helpers.OperatorClient,
+	etcdClient etcdcli.EtcdClient,
+	infrastructureLister configv1listers.InfrastructureLister,
+	eventRecorder events.Recorder,
+) factory.Controller {
+	c := &DefragController{
+		operatorClient:       operatorClient,
+		etcdClient:           etcdClient,
+		infrastructureLister: infrastructureLister,
+	}
+	return factory.New().ResyncEvery(10*time.Minute).WithInformers(
+		operatorClient.Informer(),
+	).WithSync(c.sync).ToController("DefragController", eventRecorder.WithComponentSuffix("defrag-controller"))
+}
+
+func (c *DefragController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	err := c.checkDefrag(ctx, syncCtx.Recorder())
+	if err != nil && !errors.Is(err, context.Canceled) {
+		_, _, updateErr := v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:    "DefragControllerDegraded",
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "Error",
+			Message: err.Error(),
+		}))
+		if updateErr != nil {
+			syncCtx.Recorder().Warning("DefragControllerUpdatingStatus", updateErr.Error())
+		}
+		return err
+	}
+
+	_, _, updateErr := v1helpers.UpdateStatus(c.operatorClient,
+		v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:   "DefragControllerDegraded",
+			Status: operatorv1.ConditionFalse,
+			Reason: "AsExpected",
+		}))
+	return updateErr
+}
+
+func (c *DefragController) checkDefrag(ctx context.Context, recorder events.Recorder) error {
+	controlPlaneTopology, err := ceohelpers.GetControlPlaneTopology(c.infrastructureLister)
+	if err != nil {
+		return err
+	}
+	// defrag is blocking and can only be safely performed in HighlyAvailableTopologyMode
+	if controlPlaneTopology != configv1.HighlyAvailableTopologyMode {
+		klog.V(4).Infof("DefragController is disabled for control plane topology: %s", controlPlaneTopology)
+		return nil
```

Comment:
Should we also update the operator status here to reflect this? Otherwise the status would be false degraded and AsExpected.

Reply:
good point I added a disabled status